src/server/io_utils.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/io_utils.h"
+
+#include "base/flags.h"
+#include "base/logging.h"
+#include "server/error.h"
+
+namespace dfly {
+
+void BufferedStreamerBase::ReportWritten() {
+  buffered_++;
+  // Check and possibly wait for consumer to keep up.
+  waker_.await([this]() { return !IsStalled() || IsStopped(); });
+  // Notify consumer we have new data.
+  waker_.notify();
+}
+
+std::error_code BufferedStreamerBase::WriteAll(io::Sink* dest) {
```

Comment:
`Write` part is a confusing since both the consumer and producer parts are "writing".

I needed to look into the function to understand what it does. Maybe call it `ConsumeIntoSink` or similar?